core/chains/solana/soltxm/txm.go

```diff
@@ -25,82 +34,368 @@ var (
-func NewTxm(tc func() (solanaClient.ReaderWriter, error), cfg config.Config, lggr logger.Logger) *Txm {
+func NewTxm(chainID string, tc func() (solanaClient.ReaderWriter, error), cfg config.Config, lggr logger.Logger) *Txm {
 	lggr = lggr.Named("Txm")
 	return &Txm{
 		starter: utils.StartStopOnce{},
-		tc:      tc,
 		lggr:    lggr,
-		queue:   make(chan *solanaGo.Transaction, MaxQueueLen), // queue can support 1000 pending txs
-		stop:    make(chan struct{}),
-		done:    make(chan struct{}),
+		chSend:  make(chan pendingTx, MaxQueueLen), // queue can support 1000 pending txs
+		chSim:   make(chan pendingTx, MaxQueueLen), // queue can support 1000 pending txs
+		chStop:  make(chan struct{}),
 		cfg:     cfg,
+		txs:     newPendingTxContextWithProm(chainID),
+		client:  NewLazyLoad(tc),
 	}
 }
 
 // Start subscribes to queuing channel and processes them.
 func (txm *Txm) Start(context.Context) error {
-	return txm.starter.StartOnce("solanatxm", func() error {
+	return txm.starter.StartOnce("solana_txm", func() error {
+		txm.done.Add(3) // waitgroup: tx retry, confirmer, simulator
 		go txm.run()
 		return nil
 	})
 }
 
 func (txm *Txm) run() {
-	defer close(txm.done)
-	ctx, cancel := utils.ContextFromChan(txm.stop)
+	defer txm.done.Done()
+	ctx, cancel := utils.ContextFromChan(txm.chStop)
 	defer cancel()
+
+	// start confirmer + simulator
+	go txm.confirm(ctx)
+	go txm.simulate(ctx)
+
+	for {
+		select {
+		case msg := <-txm.chSend:
+			// process tx
+			sig, err := txm.sendWithRetry(ctx, msg.tx, msg.timeout)
+			if err != nil {
+				txm.lggr.Errorw("failed to send transaction", "error", err)
+				txm.client.Reset() // clear client if tx fails immediately (potentially bad RPC)
+				continue           // skip remainining
+			}
+
+			// send tx + signature to simulation queue
+			msg.signature = sig
+			select {
+			case txm.chSim <- msg:
+			default:
+				txm.lggr.Warnw("failed to enqeue tx for simulation", "queueFull", len(txm.chSend) == MaxQueueLen, "tx", msg)
+			}
+
+			txm.lggr.Debugw("transaction sent", "signature", sig.String())
+		case <-txm.chStop:
+			return
+		}
+	}
+}
+
+func (txm *Txm) sendWithRetry(chanCtx context.Context, tx *solanaGo.Transaction, timeout time.Duration) (solanaGo.Signature, error) {
+	// fetch client
+	client, err := txm.client.Get()
+	if err != nil {
+		return solanaGo.Signature{}, errors.Wrap(err, "failed to get client in soltxm.sendWithRetry")
+	}
+
+	// create timeout context
+	ctx, cancel := context.WithTimeout(chanCtx, timeout)
+
+	// send initial tx (do not retry and exit early if fails)
+	sig, err := client.SendTx(ctx, tx)
+	if err != nil {
+		cancel()            // cancel context when exiting early
+		txm.txs.Failed(sig) // increment failed metric
+		return solanaGo.Signature{}, errors.Wrap(err, "tx failed initial transmit")
+	}
+
+	// store tx signature + cancel function
+	if err := txm.txs.Insert(sig, cancel); err != nil {
+		cancel() // cancel context when exiting early
```

Comment:
if a `defer` is used then it would also cancel the context that is used in the exponential retry when the function exits since the retry runs in a goroutine?

Reply:
Ah yes, right. Wasn't paying close attention here.